### ISSUE TITLE
Defend against malicious USB devices

### ISSUE THREAD

**Issue by brl** [subgraph/subgraph-os-issues#76]
This ticket is to collect ideas and information around the theme of hardening USB in Subgraph OS.  

Some of the attacks we want to prevent are discussed here:

https://srlabs.de/badusb/

#### Grsecurity Deny New USB
Grsecurity has a sysctl variable that can be toggled to allow or deny the connection of new USB devices:
```
kernel.grsecurity.deny_new_usb
```
This could be exposed to the user with a configuration UI of some sort, or we could use it in other ways such as automatically toggling when the screen is locked/unlocked. 

#### Linux kernel USB authorization

The Linux kernel provides a facility for authorization of new USB devices from userland:

https://www.kernel.org/doc/Documentation/usb/authorization.txt

#### USBGuard

USBGuard is a framework for creating policies for allowing or denying entire classes of devices or specific device IDs.

https://dkopecek.github.io/usbguard/

**Comment by conorsch** [comment]
Leveraging `kernel.grsecurity.deny_new_usb` is low-hanging fruit here. Mostly it protects against drive-by USB attacks, which isn't of concern to everyone, but in the context of an adversary-resistant platform, I think it makes sense.

For those playing along at home, consider enabling the option by default, then interactively whitelisting new USB devices when you need to plug something in, e.g. a hardware key token. I use an ["allow-usb" script](https://gist.github.com/conorsch/795100caf411fccf1957b3e4ff0f2843), like so:
```
$ sudo allow-usb
[sudo] password for user: 
New USB devices temporarily allowed; connect new device now.
Disabling new USB devices in 1... done.
```

The script will take care of reenabling the blacklist, which is far more usable interactiving than futzing with sysctl options back and forth.

**Comment by dma** [comment]
Hey Conor,

We've (@xSmurf) implemented a related PoC called usblockout. It is tied to desktop locking - when a desktop lock occurs, deny_new_usb is toggled on. It's in our repo but we haven't made it default. You can try it with:

sudo apt-get install usblockout

Code is here:
 https://github.com/subgraph/usblockout

**Comment by conorsch** [comment]
@dma Magnificent, this looks great. Love that it can be used interactively so well. Trying it out!

**Comment by dma** [comment]
@conorsch yes, combined with our (soon to come) event monitor/notifier, it's quite nice - screenshot: https://twitter.com/attractr/status/777220653447581701
